content/blog/ebpf-openssl-tracing/index.md

```diff
@@ -0,0 +1,340 @@
+
+1. Give Wireshare the RSA private key, so that it can trace the entire connection. In this setup, Wireshark is able to trace the data as long as it was able to observe the initial key-exchange; for existing connections, you’ll be out of luck.
+2. Give Wireshark the (pre)-master key which was negotiated during the key exchange—if you know it. In this setup, Wireshark can decrypt the data even without seeing the initial connection. Because this is a common use case, you can actually ask your [browser to dump the session keys](https://redflagsecurity.net/2019/03/10/decrypting-tls-wireshark/) for this purpose.
+
+Sharing encryption keys works well for Wireshark’s use case, but this approach is not scalable enough to use in an observability platform whose mantra is “up and running in seconds.”
+
+## A simple solution
+
+Given how powerful BPF is, we wondered if there wasn’t a more seamless way to access the plain-text data. The resulting plan is very basic: we simply capture the data before it is encrypted.
+
+This pre-encryption approach rules out kprobes, because the data is encrypted before it reaches the Linux kernel. But BPF isn’t restricted to kernel event triggers. Uprobes allow the kernel to trigger a BPF program whenever the *application code* reaches a certain instruction.
+
+To figure out where to place the uprobe in order to capture the pre-encryption traffic, it helps to look at a diagram of how SSL/TLS is typically incorporated in applications.
+
+::: div image-xl
+<svg title='Diagram of application without TLS vs Application with TLS.' src='bpf-tls-tracing.svg' />
+:::
+
+The diagram above shows the common setup for encrypting application traffic with the popular TLS library, OpenSSL. The goal is to trace the inputs to the TLS library, which happen to be the `SSL_write` and `SSL_read` functions in OpenSSL. Tracing these calls, instead of the Linux `send()` and `recv()` syscalls, will capture the traffic before it is encrypted.
+
+This strategy is well supported by BPF *uprobes*. With uprobes, we set our triggers to be events that happen in user-space. Often, uprobes are set in the user’s own compiled code, but there’s nothing preventing us from placing them on a shared library.
+
+By placing the uprobe on a *shared* library, we end up tracing *all* applications that use the shared library, so we’ll need to filter out the data for the processes we’re actually interested in. This is no different than when we put a kprobe on Linux’s `send()` and `recv()` syscalls, though. Those kprobes also end up tracing *all* applications —- in fact, they trace more applications, as not all applications will use OpenSSL, but all applications will go through the kernel. In the context of a full system tracer, however, a probe on *shared* library is actually an advantage, since a single probe gives us wide observability coverage.
```

Comment:
not all applications will use OpenSSL -> not all applications use OpenSSL
all applications will go through the kernel -> all aplications go through the kernel